service/src/main/kotlin/fi/espoo/evaka/invoicing/domain/FeeDecisions.kt

```diff
@@ -148,6 +140,42 @@ enum class FeeDecisionType {
     RELIEF_ACCEPTED
 }
 
+@ConstList("feeDecisionDifferences")
+enum class FeeDecisionDifference(val contentEquals: (d1: FeeDecision, d2: FeeDecision) -> Boolean) : DatabaseEnum {
+    GUARDIANS({ d1, d2 -> setOf(d1.headOfFamilyId, d1.partnerId) == setOf(d2.headOfFamilyId, d2.partnerId) }),
+    CHILDREN({ d1, d2 -> d1.children.map { it.child.id }.toSet() == d2.children.map { it.child.id }.toSet() }),
+    INCOME({ d1, d2 ->
+        setOf(d1.headOfFamilyIncome, d1.partnerIncome) == setOf(
+            d2.headOfFamilyIncome,
+            d2.partnerIncome
+        ) && childrenEquals(d1, d2) { it.childIncome }
+    }),
+    PLACEMENT({ d1, d2 -> childrenEquals(d1, d2) { it.placement } }),
+    SERVICE_NEED({ d1, d2 -> childrenEquals(d1, d2) { it.serviceNeed } }),
+    SIBLING_DISCOUNT({ d1, d2 -> childrenEquals(d1, d2) { it.siblingDiscount } }),
+    FEE_ALTERATIONS({ d1, d2 -> childrenEquals(d1, d2) { it.feeAlterations } }),
+    FAMILY_SIZE({ d1, d2 -> d1.familySize == d2.familySize }),
+    FEE_THRESHOLDS({ d1, d2 -> d1.feeThresholds == d2.feeThresholds });
+
+    override val sqlType: String = "fee_decision_difference"
+
+    companion object {
+        fun getDifference(d1: FeeDecision, d2: FeeDecision): Set<FeeDecisionDifference> {
+            if (d1.isEmpty() && d2.isEmpty()) {
+                return if (GUARDIANS.contentEquals(d1, d2)) emptySet() else setOf(GUARDIANS)
+            }
+            return values().filterNot { it.contentEquals(d1, d2) }.toSet()
+        }
+
+        private fun <T> childrenEquals(d1: FeeDecision, d2: FeeDecision, fn: (child: FeeDecisionChild) -> T): Boolean {
+            val map1 = d1.children.associateBy({ it.child.id }, fn)
+            val map2 = d2.children.associateBy({ it.child.id }, fn)
+            // equals only with matching keys
+            return map1.all { (key, value) -> map2.getOrDefault(key, value) == value }
```

Comment:
I think this should be just a normal equality check, ie. `map1 == map2`.

Reply:
That would mean every child change on decision there would be difference with income, placement, service need, sibling discount and fee alterations. Because of that it was said "in the spec" that it should show only `CHILDREN` difference when a child is added/removed (if there are no other changes of course, well sibling discount can be different by just child added/removed, but it already works with current implementation).

Reply:
Ah yes, now I get it!